core/drivers/crypto/crypto_api/cipher/cipher.c

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2020 NXP
+ *
+ * Crypto Cipher interface implementation to enable HW driver.
+ */
+#include <assert.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+#include <drvcrypt.h>
+#include <drvcrypt_cipher.h>
+#include <kernel/panic.h>
+#include <malloc.h>
+#include <utee_defines.h>
+#include <util.h>
+
+static const struct crypto_cipher_ops cipher_ops;
+
+/*
+ * Returns the reference to the driver context
+ *
+ * @ctx    Reference the API context pointer
+ */
+static struct crypto_cipher *to_cipher_ctx(struct crypto_cipher_ctx *ctx)
+{
+	assert(ctx && ctx->ops == &cipher_ops);
+
+	return container_of(ctx, struct crypto_cipher, cipher_ctx);
+}
+
+/*
+ * Free cipher context
+ *
+ * @ctx    Reference the API context pointer
+ */
+static void cipher_free_ctx(struct crypto_cipher_ctx *ctx)
+{
+	struct crypto_cipher *cipher = to_cipher_ctx(ctx);
+
+	if (cipher->op && cipher->op->free_ctx)
+		cipher->op->free_ctx(cipher->ctx);
+
+	free(cipher);
+}
+
+/*
+ * Copy cipher context
+ *
+ * @dst_ctx  [out] Reference the API context pointer destination
+ * @src_ctx  Reference the API context pointer source
+ */
+static void cipher_copy_state(struct crypto_cipher_ctx *dst_ctx,
+			      struct crypto_cipher_ctx *src_ctx)
+{
+	struct crypto_cipher *cipher_src = to_cipher_ctx(src_ctx);
+	struct crypto_cipher *cipher_dst = to_cipher_ctx(dst_ctx);
+
+	if (cipher_src->op && cipher_src->op->copy_state)
+		cipher_src->op->copy_state(cipher_dst->ctx, cipher_src->ctx);
+}
+
+/*
+ * Initialization of the cipher operation
+ *
+ * @ctx      Reference the API context pointer
+ * @mode     Operation mode
+ * @key1     First Key
+ * @key1_len Length of the first key
+ * @key2     Second Key
+ * @key2_len Length of the second key
+ * @iv       Initial Vector
+ * @iv_len   Length of the IV
+ */
+static TEE_Result cipher_init(struct crypto_cipher_ctx *ctx,
+			      TEE_OperationMode mode, const uint8_t *key1,
+			      size_t key1_len, const uint8_t *key2,
+			      size_t key2_len, const uint8_t *iv, size_t iv_len)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	struct crypto_cipher *cipher = to_cipher_ctx(ctx);
+
+	if ((!key1 && key1_len) || (!key2 && key2_len) || (!iv && iv_len)) {
+		CRYPTO_TRACE("One of the key is not correct");
+		CRYPTO_TRACE("key1 @%p-%zu bytes", key1, key1_len);
+		CRYPTO_TRACE("key2 @%p-%zu bytes", key1, key1_len);
+		CRYPTO_TRACE("iv   @%p-%zu bytes", iv, iv_len);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (cipher->op && cipher->op->init) {
+		struct drvcrypt_cipher_init dinit = {
+			.ctx = cipher->ctx;
```

Comment:
s/`;`/`,`/g in the structure instance.

Reply:
Oups, my bad for this one. I will check on the test failure.